gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMergerTest.groovy

```diff
@@ -0,0 +1,88 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.dist
+
+import org.codehaus.groovy.runtime.InvokerHelper
+import spock.lang.Specification
+
+class RecommendedProductDependencyMergerTest extends Specification {
+    def basicDep = new RecommendedProductDependency(
+        productGroup: "group",
+        productName: "name")
+
+    def "picks larger minimum version and smaller maximum version"() {
+        given:
+        def dep1 = newRecommendation("2.0.0", "2.6.x", "2.2.0")
+        def dep2 = newRecommendation("2.1.0", "2.x.x", "2.2.0")
+
+        when:
+        def merged = RecommendedProductDependencyMerger.mergeRecommendedProductDependencies(dep1, dep2)
+
+        then:
+        merged.minimumVersion == "2.1.0"
+        merged.maximumVersion == "2.6.x"
+        merged.recommendedVersion == "2.2.0"
+    }
+
+    def "fails if new min is greater than new max"() {
+        given:
+        def dep1 = newRecommendation("2.0.0", "2.6.x", "2.2.0")
+        def dep2 = newRecommendation("2.7.0", "2.x.x", "2.8.0")
+
+        when:
+        def merged = RecommendedProductDependencyMerger.mergeRecommendedProductDependencies(dep1, dep2)
+
+        then:
+        def e = thrown(IllegalArgumentException)
+        e.message.contains("Inferred minimum version does not match inferred maximum version")
+    }
+
+    def "fails if min == max"() {
+        given:
+        def dep1 = newRecommendation("2.5.0", "2.x.x", null)
+        def dep2 = newRecommendation("2.1.0", "2.5.0", null)
+
+        when:
+        def merged = RecommendedProductDependencyMerger.mergeRecommendedProductDependencies(dep1, dep2)
+
+        then:
+        def e = thrown(IllegalArgumentException)
+        e.message.contains("minimumVersion and maximumVersion must be different")
+    }
+
+    def "max and recommended can be optional"() {
+        given:
+        def dep1 = newRecommendation("2.1.0", null, "2.6.0")
+        def dep2 = newRecommendation("2.1.0", "2.8.x", null)
+
+        when:
+        def merged = RecommendedProductDependencyMerger.mergeRecommendedProductDependencies(dep1, dep2)
+
+        then:
+        merged.minimumVersion == "2.1.0"
+        merged.maximumVersion == "2.8.x"
+        merged.recommendedVersion == "2.6.0"
+    }
+
+    private RecommendedProductDependency newRecommendation(String min, String max, String recommended) {
+        def rpd = basicDep.clone() as RecommendedProductDependency
+        use InvokerHelper, {
```

Comment:
woah. Is this really how we construct RecommendedProductDependencies elsewhere??

Reply:
no, you can just do `RecommendedProductDependency(an inline map)`
In the gradle plugin the extension gives you a configuration point that takes a Closure: https://github.com/palantir/sls-packaging/blob/develop/gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesExtension.groovy#L25

Reply:
This is just so I can clone things so I don't have to re-type the same fields that never change... I didn't know a better way to do that in groovy